### PR TITLE
Signtool perf improvements

### PR DESCRIPTION
**Pull request by mmitche**
This drastically improves the performance of post build signing, or really any large scale signing.
- Rework Configuration to reduce the number of times we compute the hash of a file. Prior to this change, this operation accounts for ~80-90% of the total signtool time.
- Repack containers in parallel
- Remove forceRepack switch (unused)
- Remove container cache (unused)

### CODE REVIEW

**Comment by mmitche** [comment]
Before: 123.29 mins
After: 33.44 mins

72% improvement

File: src/Microsoft.DotNet.SignTool.Tests/Microsoft.DotNet.SignTool.Tests.csproj [hunk 1]
```diff
@@ -1,18 +1,15 @@
 <!-- Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. -->
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>net472;netcoreapp2.1</TargetFrameworks>
```
Comment (chcosta) [inline src/Microsoft.DotNet.SignTool.Tests/Microsoft.DotNet.SignTool.Tests.csproj:4]: Are the changes to this file intentional or just a merge issue?

File: src/Microsoft.DotNet.SignTool/src/Configuration.cs [hunk 2]
```diff
@@ -224,16 +185,17 @@ internal BatchSignInput GenerateListOfFiles()
             string fullPath,
             string collisionPriorityId,
             ImmutableArray<byte> contentHash,
-            bool isNested,
-            bool forceRepack = false,
```
Comment (chcosta) [inline src/Microsoft.DotNet.SignTool/src/Configuration.cs:228]: Should probably also remove ForceRepack from FileSignInfo

File: src/Microsoft.DotNet.SignTool/src/Configuration.cs [hunk 3]
```diff
@@ -673,51 +643,52 @@ private bool TryBuildZipData(FileSignInfo zipFileSignInfo, string containerPath,
...
+                                using (var tempFileStream = File.OpenWrite(tempPath))
+                                {
+                                    entryMemoryStream.CopyTo(tempFileStream);
+                                }
 
-                            _hashToCollisionIdMap.TryGetValue(fileUniqueKey, out string collisionPriorityId);
-                            fileSignInfo = TrackFile(tempPath, collisionPriorityId, contentHash, isNested: true, containerPath: containerPath ?? archivePath);
-                        }
+                                _hashToCollisionIdMap.TryGetValue(fileUniqueKey, out string collisionPriorityId);
+                                fileSignInfo = TrackFile(tempPath, collisionPriorityId, contentHash, parentContainerPath: containerPath, parentContainerHash: containerHash);
```
Comment (chcosta) [inline src/Microsoft.DotNet.SignTool/src/Configuration.cs:681]: we keep passing container path and hash around everywhere.  Does it make sense to make it an immutable object that can be passed around?

**Review by chcosta — COMMENTED** [review]
This looks really nice.  a few minor comments.

File: src/Microsoft.DotNet.SignTool/src/Configuration.cs (hunk 3 above)
Comment (mmitche) [reply src/Microsoft.DotNet.SignTool/src/Configuration.cs:681]: You mean combine them together into an immutable object?

Comment (chcosta) [reply src/Microsoft.DotNet.SignTool/src/Configuration.cs:681]: yes

File: src/Microsoft.DotNet.SignTool.Tests/Microsoft.DotNet.SignTool.Tests.csproj (hunk 1 above)
Comment (mmitche) [reply src/Microsoft.DotNet.SignTool.Tests/Microsoft.DotNet.SignTool.Tests.csproj:4]: Not intentional.

**Comment by mmitche** [comment]
@chcosta Created a new immutable type for path+hash

File: src/Microsoft.DotNet.SignTool/src/Configuration.cs [hunk 4]
```diff
@@ -633,6 +630,7 @@ private bool TryBuildZipData(FileSignInfo zipFileSignInfo, string containerPath,
                 {
                     var nestedParts = new Dictionary<string, ZipPart>();
 
+
```
Comment (chcosta) [inline src/Microsoft.DotNet.SignTool/src/Configuration.cs:633]: nit: extra space

**Review by chcosta — APPROVED** [review]
The immutable object change was a lot larger than I had imagined, sorry about that.  This looks good though!

**Pull request merged**